stdlib/public/core/AutoDiff.swift

```diff
@@ -183,6 +183,41 @@ public extension Differentiable where TangentVector == Self {
   }
 }
 
+/// A type that consists of a differentiable vector space and some other
+/// non-differentiable component.
+///
+/// Mathematically, this represents a product manifold that consists of
+/// a differentiable vector space and some arbitrary manifold, where the tangent
```

Comment:
It seems non-controversial that types are isomorphic to sets/spaces.

Wikipedia suggests manifolds are special topological spaces:
![image](https://user-images.githubusercontent.com/5590046/63646610-4b44bc00-c6ca-11e9-9a2d-9eda9aaaf36b.png)

Though this description seems to imply that all types can be interpreted as "`some arbitrary manifold`", including `Bool` and other wildly "non-differentiable" types. Is the description accurate?

Reply:
Maybe rephrasing "some arbitrary manifold" as "some other arbitrary component" would be less controversial, as "component" is vague.

Reply:
We've had the assumption already for the base `Differentiable` protocol, which allows us to include values of arbitrary types as stored properties and to call the whole type a "differentiable manifold". This should definitely be addressed if brought up in a Swift evolution review.

Reply:
> Maybe rephrasing "some arbitrary manifold" as "some other arbitrary component" would be less controversial, as "component" is vague.

IIUC, the product of a manifold and a non-manifold is no longer a manifold. This would contradict even the base protocol where everything `Differentiable` is a differentiable manifold (hence a manifold).

Reply:
> We've had the assumption already for the base Differentiable protocol, which allows us to include values of arbitrary types as stored properties and to call the whole type a "differentiable manifold".

That's a great point, thanks 🙂